plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,455 @@
+
+We use a different approach where we prior clean the AST PlutusIR term from all the lets (rec&nonrec) and store them in a separate 'RhsTable'.
+We use the information obtained by the dependency graph to create let-(rec&nonrec) groups at that rank. A rank is a lambda depth plus a lambda name. During
+the second pass, whenever we reach a lambda/Lambda definition we float all lets beloning to this lambda (rank, i.e. lambda's depth and name).
+This means that the lets are floated *exactly inside a lambda definition, i.e at the lambda body*, with the downside being that they are floated "a bit" more outside than what would
+be optimal as in the original algorithm (floated exactly around the 'expr' that needs them, and *exactly outside a lambda definition*).
+The upside is that this approach is not local/ad-hoc, but more holistic.
+
+-}
+
+
+-- | During the first-pass we compute a maximum "rank" for every let-binding declared in the given PIR program.
+-- See Note: [Float first-pass]
+-- A rank points to either a lambda/Lambda location or Top if the let does not depend on any lambda/Lambda.
+-- A "maximum rank" is the 'Rank' with the largest depth, i.e. the deepest lambda-dependency.
+-- A rank can also be used as a pointer to a (lambda) location in the PIR program.
+data Rank = Top -- ^ signifies that a let has no lambda/Lambda dependency and thus can be placed at the toplevel of the program
+          | LamDep Int PLC.Unique -- ^ a let has a lambda/Lambda dependency with lamDepth :: 'Int', lamName :: 'PLC.Unique'
+          deriving Eq
+
+-- | The underlying classic getter function for depth
+--
+-- Note: Top is arbitrarily defined as having depth 0.
+-- It could also be made 'minBound' or 'Word', but this is conceptually clearer and allows us to use 'IM.IntMap's.
+_depth :: Rank -> Int
+_depth = \case
+  Top -> 0
+  LamDep d _ -> d
+
+-- | Lens-style getter of the depth of a rank
+depth :: Getting r Rank Int
+depth  = to _depth
+
+instance Ord Rank where
+  compare = compare `on` _depth
+
+-- | We are interested in maximal ranks, and thus we use Rank similar to 'Data.Semigroup.Max'.
+-- OPTIMIZE: we could get rid of the following Semigroup and Monoid instances by wrapping with 'Data.Semigroup.Max'.
+instance Semigroup Rank where
+  r1 <> r2 = max r1 r2
+
+-- | The minimum rank is Top (toplevel/ no lambda/Lambda dependency).
+instance Monoid Rank where
+  mempty = Top
+
+-- | During the first pass of the AST, a reader context holds the current in-scope lambda locations (as Ranks).
+type P1Ctx = ( P1Data  --  the lambdas-in-scope that surround an expression
+             , Int     --  the current depth
+             )
+
+
+-- | During the first pass of the AST, we build an intermediate table to hold the maximum ranks of each let encountered so far.
+-- This intermediate table will be transformed right at the end of 1st pass to 'FloatData'.
+type P1Data = M.Map
+              PLC.Unique --  the identifier introduced by a let-binding
```

Comment:
`M.Map PLC.Unique` is better known as `Language.PlutusCore.Name.UniqueMap PLC.Unique`

Reply:
Although in this case it's probably even less convenient (still faster though).

Reply:
Yes, less convenient I keep it as it is for the moment.